src/EhsnPlugin/Mappers/ReadingMapper.cs

```diff
@@ -203,62 +203,79 @@ private void AddGageHeightReading(List<Reading> readings, EHSNStageMeasStageMeas
             var srcAction = stageMeasurement.SRCApp;
-
-            AddLoggerReading(readings, time, hg1, src, srcAction);
-            AddLoggerReading(readings, time, hg2, src, srcAction);
-            AddWaterLevelReading(readings, time, wl1, timeText, hg1, hg2, src, srcAction);
-            AddWaterLevelReading(readings, time, wl2, timeText, hg1, hg2, src, srcAction);
+            var readingType = stageMeasurement.ReadingType;
+            var wl1Header = SanitizeBenchmarkName(_eHsn?.StageMeas?.WL1Header);
+            var wl2Header = SanitizeBenchmarkName(_eHsn?.StageMeas?.WL2Header);
+            var gc1 = _eHsn?.StageMeas?.GCWL1.ToNullableDouble();
+            var gc2 = _eHsn?.StageMeas?.GCWL2.ToNullableDouble();
+            string hg1Header = "WL Source: HG";
+            string hg2Header = "WL Source: HG2";
+
+            AddLoggerReading(readings, time, hg1, hg1Header, src, srcAction);
+            AddLoggerReading(readings, time, hg2, hg2Header, src, srcAction);
+            AddWaterLevelReading(readings, time, wl1, gc1, wl1Header, hg1, hg2, src, srcAction, readingType);
+            AddWaterLevelReading(readings, time, wl2, gc2, wl2Header, hg1, hg2, src, srcAction, readingType);
         }
 
-        private void AddWaterLevelReading(List<Reading> readings, DateTimeOffset time, double? wl,
-            string timeText, double? hg1, double? hg2, double? src, string srcAction)
+        private void AddWaterLevelReading(List<Reading> readings, DateTimeOffset time, double? wl, double? gc, string wlHeader,
+             double? hg1, double? hg2, double? src, string srcAction, string readingType)
         {
             if (!wl.HasValue) return;
 
-            var referencePointName = GetReferencePointName(timeText);
+            var hgComments = GetHgComment(gc, hg1, hg2);
 
-            var hgComments = GetHgComment(hg1, hg2);
+            if (gc.HasValue)
+            {
+                wl += gc;
+            }
 
             var reading = AddReading(readings, time, Parameters.StageHg, Units.DistanceUnitId, wl.ToString());
 
             reading.Method = Config.StageWaterLevelMethodCode;
-            reading.ReferencePointName = referencePointName;
+            reading.ReferencePointName = wlHeader;
             reading.Publish = true;
-            reading.ReadingType = "Reset (RS)".Equals(srcAction, StringComparison.InvariantCultureIgnoreCase)
-                ? ReadingType.ResetAfter
-                : string.IsNullOrWhiteSpace(srcAction)
-                    ? ReadingType.Routine
-                    : ReadingType.Unknown;
+            if (string.IsNullOrWhiteSpace(readingType))
+            {
+                reading.ReadingType = ReadingType.Unknown;
+            }
+            else
+            {
+                string formattedReadingType = readingType.Replace(" ", String.Empty);
+                reading.ReadingType = (ReadingType)Enum.Parse(typeof(ReadingType), formattedReadingType);
+            }
```

Comment:
You can also use the Enum.TryParse<TEnum> method for more robust enum parsing.

This whole if-then-else block becomes this line.

```C#
reading.ReadingType = Enum.TryParse<ReadingType>(readingType, true, out var value) ? value : ReadingType.Unknown;
```
That will trim any leading/trailing whitespace, parse the enum value case-insensitively and fallback to Unknown when the parse fails.